### DISCUSSION TITLE
Cart Transform API Limitations/Issues/Opportunities (Customer Account Order | Notifications | Liquid))

### DISCUSSION THREAD

**Discussion by dnpg** [Shopify/function-examples#541] [Feedback: Cart Transform API]
The intention of this discussion is to show the current opportunities/issues/limitations I found when using the Cart Transform API to create <b>Customized Bundles</b>.

Just to clarify, when I say "<b>Customized Bundle</b>", I'm talking about the bundles created with Shopify Functions when executing Merge/Expand Operations in the Cart.

I'm going to try to be as clear as possible, so that is why I divided my comments in sections below:

<h2>Admin Orders</h2>
<br/>
<ol>
<li>🔒Limitation: All line items are displayed individually and not grouped as per checkout.</li>
<li>🔒Limitation: The order line items seems to display the Variant title instead of the CartLine title. (With the new Cart Transform API -> Update operation we can change this to whatever we want)</li>
</ol>
<br/>
<b>Example of the current Admin view:</b>
<br/>
<br/>
<img width="400" alt="Screenshot 2024-07-20 at 5 54 56 PM" src="https://github.com/user-attachments/assets/05368b24-0916-4a4e-bb26-44ca4bc1b1fe"/><br/><em>Example of an order in the Shopify Admin</em>
<br/><br/>
✅ <b>Expected:</b><br/>
<ol>
<li> Display the Order Summary items in the New Customer Accounts Order the same way as per Checkout and Thank You Page. <br/><br/>
<img width="1180" alt="Screenshot 2024-07-20 at 3 12 40 PM" src="https://github.com/user-attachments/assets/0520b4c6-6542-4d8f-8f80-de13a07a91e4"/><br/><em>Example of an order in the Thank You Page</em><br/><br/>
💡<b>Suggestions:</b>  Maybe include a toggle button to view just the individual line items VS the bundled item as per Checkout (Which is what the customer viewed).</li>
<li>Display the actual line item as it was displayed in the Checkout instead of the Variant title. Maybe have the option to display the variant title on hover</li>
</ol>

<h2>New Customer Accounts VS Order page</h2>
<p>Below is an example of a <b>Thank You Page</b> view that contains a product where an <b>expand operation</b> was executed. This is very clear for the customer:<br/><br/>
<img width="1180" alt="Screenshot 2024-07-20 at 3 12 40 PM" src="https://github.com/user-attachments/assets/0520b4c6-6542-4d8f-8f80-de13a07a91e4"/><br/>
<em>The Order Summary at Checkout shows the same visual</em>.<br/>
<br/>If the customer refreshes the page, they are redirected to the <b>New Customers Account Order</b> (Even if this has not being enabled in the store, or at least this happens on my dev store) page. And in this view, this is what the customer sees:<br/><br/>
<img width="1187" alt="Screenshot 2024-07-20 at 3 13 01 PM" src="https://github.com/user-attachments/assets/6a07d236-d519-4a0f-baee-cae21bcc388e"/><br/><em>New Customer Accounts Order  View</em>
<br/>
<h3>Limitations</h3>
<ol>
<li>The <b>customer</b> sees a confusing view of all the individual items instead of the grouped view that they are familiar with from the Checkout and Thank You Pages.</li>
<li>Some <b>merchants</b> don't want the customers to see the individual pricing of the<b>"bundled items"</b>.</li>
<li>The <b>New Customer Accounts Order</b> view doesn't have any extensions available to modify/hide how the products are displayed</li>
</ol>
<br/>
✅ <b>Expected behaviour</b>:Display the Order Summary items in the New Customer Accounts Order the same way as per Checkout and Thank You Page.
<br/><br/>
💡 <b>Suggestions</b>:  If the current behaviour might be needed for some merchants, then allow an additional setting in the Customers Account Admin to enable or disable the option to Group the <b>"bundled items"</b> in the Order view.
</p>

<h2>GraphQL / Rest / Theme Liquid / Notifications Liquid / Inconsistencies</h2>
<br/>
<b>Theme Liquid VS Notifications Liquid VS GraphQL properties</b>: 
<br/><br/>The <b>Customer Order Confirmation</b> notification liquid code (by Shopify) has the following code:<br/>
<img width="675" alt="Screenshot 2024-07-20 at 4 45 02 PM" src="https://github.com/user-attachments/assets/06242472-bc25-48f2-a960-142b62dba303"/><em>Order Confirmation line.group in use</em>.<br/><br/> That code is the one displaying the  "Part of [Bundle product title]" under line items when the item is part of a bundle.
<br/>
Based on that, I found the follwing:<br/><br/>
<ol>
<li><b>line.group</b> is not available in the Theme Liquid <b>line_item (As part of the Order object)</b> object  <a href="https://shopify.dev/docs/api/liquid/objects/line_item#line_item-item_components" target="_blank">line_item Liquid objects here</a>. Hence, we are not able to replicate the same descriptive view in the New Customer Account Order view and the templates/customers/order.liquid</li>
<li>The line.group liquid "line" property is 🚫📄 not officially documented (<a href="https://help.shopify.com/en/manual/orders/notifications/email-variables#line-item" target="_blank">See documentation here</a>).
</li>
<li>GraphQL Cart API seems to miss the CartLine <b>title</b> field. This field when used in liquid, not only contains historical data on the name of the Variant/Product when purchased, it also reflect changes made with the new Cart Transform API -> Update operation.
<br/><br/>
<img width="593" alt="Screenshot 2024-07-22 at 1 21 53 PM" src="https://github.com/user-attachments/assets/59490791-349d-419d-854d-f13c51423d79">
</li>
</ol>
<br/><br/>
📝<b>Note 1</b>: GraphQL does have a similar object as per Notifications Liquid called LineGroup, see offical docs <a href="https://shopify.dev/docs/api/admin-graphql/2024-07/objects/LineItem#field-lineitemgroup" target="_blank">here</a>.
<br/><br/>
📝<b>Note 2</b>: We do have access to line_items.item_components in liquid (Only when in Cart object), which is similar to the <a href="https://shopify.dev/docs/api/storefront/2024-07/objects/ComponentizableCartLine#field-linecomponents">lineComponents field in the ComponentizableCartLine Object</a>
<br/><br/>
✅<b>Expected behaviour</b>: line_item in the Theme Liquid should have access to the group property so we can replicate similar functionality in the <b>templates/customers/order.liquid</b> file

<h2>Storefront / Liquid / Theme files</h2>
<br/>
From the above points, you can see in the Order Confirmation Notification Email and New Customers Account Order view, the customer is able, at least, to see a "Part of: [Bundle product]" text that gives a little bit more information to where that item is coming from. <br/>However, In the template/customers/order.liquid file we are not able to replicate that behaviour.<br/><br/>
<ol>
<li>There is no <b>line_item.group</b> property in the Theme Liquid line_item object (<a href="https://shopify.dev/docs/api/liquid/objects/line_item" target="_blank">See the official docs</a>), where line.group is actually available in the Order Confirmation Notification Liquid template (🚫📄 But not documented. <a href="https://help.shopify.com/en/manual/orders/notifications/email-variables#line-item" target="_blank">See notifications docs</a>)<br/><br/>Without the line_item.group we are not able to replicate in Liquid the same visuals in the New Customer Accounts Order view or the Order Confirmation Notification email.</li>
</ol>

<h2>Improve Documentation</h2>
<br/>
It would be great to improve some of the documentation in Shopify in relation to Bundles.
<br/><br/>
<ol>
<li>Add more clarification on <a href="https://shopify.dev/docs/api/liquid/objects/line_item#line_item-item_components" target="_blank">line_item.item_components</a>. And include a real life example on how/when to use it</li>
<li>As it has been deployed and is in use the <a href="https://help.shopify.com/en/manual/orders/notifications/email-variables#line-item" target="_blank">line.group</a> property, this should be added to the official <a href="https://help.shopify.com/en/manual/orders/notifications/email-variables#line-item" target="_blank">Notifications Liquid documentation</a>.</li>
<li>Explain a bit more the reason behind the "Part of: [Bundle product]" in the Bundles documentation, cart notifications liquid and anywhere else where is displayed</li>
<li>Clearer information in regards Inventory management and the different Cart Transform Operations. 
<br/><br/><b>For example:</b> It seems that when you run an <b>expand</b> cart transform, the <b>cartLineId</b> where the expand operation is executed is not added as a line item to the cart. When the checkout is completed and you review the items.
<br/><br/>
</li>
<li>Add documentation to respond to the question. How can I get a report on how many "Fixed Bundles" and "Customized Bundles" have been purchased?<br/><br/>
📝<b>Note</b>:  I have tried to get this and failed. When I click on the "Part of [Bundle product]" in the Admin order line item, I get redirected to the bundle product page. But If the product came from a "Customized Bundle", the inventory is not updated. Now for "Fixed bundles", or at least the ones I tested using the official Bundle app by Shopify, this one does show and Inventory but is based on the components inventory which is  not the case for "Customized Bundles".
</li>
</ol>

**Comment by paul-muckypuddle** [comment]
We're also hitting problems related to this issue - if a customer has multiple bundles in their cart, there's no way to discern what lines belong to which bundle unless we add own own line item property on 'add'.

**Comment by andrewchen5678** [comment]
There is an issue with the cart transform function today: it shows the line item attributes on the original line item before transformation but goes away after checkout since that original line item before transformation with the attributes gets removed after checkout:

https://community.shopify.com/c/shopify-functions/cart-transform-expand-operation-does-not-persist-attributes/td-p/2502517

the checkout page should hide all line item attributes of the original parent line item if they are expected to go away after checkout.

**Comment by andrewchen5678** [comment]
Is there also any easy way to show the total breakdown of the subitems inside of the bundle with checkout extensibility? I tried to do that but there is only `purchase.checkout.cart-line-item.render-after` target which only enables showing pricing info after the entire bundle rather than right next to each individual subitem.

**Comment by devenini** [comment]
Has anyone found a workaround to "group" items on the Admin Order page instead of displaying all line items individually?
